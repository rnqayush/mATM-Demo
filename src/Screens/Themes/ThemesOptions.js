import React from 'react';
import { Appearance } from 'react-native-appearance';
import PropTypes from 'prop-types';

import { Text } from '@idfc/ccl-mobile';
import LightTheme from '@idfc/ccl-commons/assets/images/light-mode.svg';
import DarkTheme from '@idfc/ccl-commons/assets/images/dark-mode.svg';
import SystemTheme from '@idfc/ccl-commons/assets/images/system-mode.svg';

import Labels from 'translations/settingsScreen/labels';

import { Wrapper, Column, StyledRadioButton } from './ThemesSettings.style';
import { THEME_NAMES } from './ThemesConstants';

const NOOP = () => {};

const ThemesOptions = ({ selectedThemeOption, onThemeChange }) => {
  const isThemeSwitchSupportedByOS = Appearance.getColorScheme() !== 'no-preference';

  const hideSystemMode = !isThemeSwitchSupportedByOS;

  const themesData = [
    {
      id: 0,
      svgComponent: <LightTheme width={79} height={170} />,
      label: Labels.LIGHT,
      value: THEME_NAMES.LIGHT,
    },
    {
      id: 1,
      svgComponent: <DarkTheme width={79} height={170} />,
      label: Labels.DARK,
      value: THEME_NAMES.DARK,
    },
    {
      id: 2,
      svgComponent: <SystemTheme width={79} height={170} />,
      label: Labels.SYSTEM,
      value: THEME_NAMES.SYSTEM,
    },
  ];

  return (
    <Wrapper>
      {themesData.map(theme => {
        const { svgComponent, id, value, label } = theme;
        return label === Labels.SYSTEM && hideSystemMode ? null : (
          <Column key={`${id}-${label}`}>
            {svgComponent}
            <StyledRadioButton
              id={id}
              value={value}
              testID={`${id}-${value}`}
              checked={selectedThemeOption === value}
              onChange={() => onThemeChange(value)}
            />
            <Text align="center" fontSize={16} lineHeight={24}>
              {label}
            </Text>
          </Column>
        );
      })}
    </Wrapper>
  );
};

export default ThemesOptions;

ThemesOptions.propTypes = {
  selectedThemeOption: PropTypes.string,
  onThemeChange: PropTypes.func,
};

ThemesOptions.defaultProps = {
  selectedThemeOption: THEME_NAMES.LIGHT,
  onThemeChange: NOOP,
};
