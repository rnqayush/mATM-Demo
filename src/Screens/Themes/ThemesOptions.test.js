import React from 'react';
import { View as MockView } from 'react-native';
import { Appearance } from 'react-native-appearance';
import { render, fireEvent } from 'test/test-utils';
import ThemesOptions from './ThemesOptions';

jest.mock('react-native-appearance', () => ({
  Appearance: {
    getColorScheme: jest.fn().mockReturnValue('light'),
    addChangeListener: jest.fn().mockReturnValue({ remove: jest.fn() }),
  },
  AppearanceProvider: ({ children }) => <MockView>{children}</MockView>,
}));

jest.mock('@idfc/ccl-mobile', () => {
  const RealModule = jest.requireActual('test/ccl-mobile.mock');
  const RenderOnlyProps = props => <MockView {...props} />;
  return {
    ...RealModule,
    RadioButton: RenderOnlyProps,
    Text: RenderOnlyProps,
  };
});

const mockThemeChange = jest.fn();
const props = { selectedThemeOption: 'LIGHT', onThemeChange: mockThemeChange };

describe('ThemeSettings screen', () => {
  it('should render component', () => {
    const tree = render(<ThemesOptions {...props} />);

    expect(tree).toMatchSnapshot();
  });

  it('should check light theme radio button by default', () => {
    const container = render(<ThemesOptions />);
    const { getByTestId } = container;

    expect(getByTestId('0-LIGHT').props.checked).toStrictEqual(true);
  });

  it('should check dark theme radio button when selectedThemeOption prop is dark', () => {
    const container = render(<ThemesOptions {...props} selectedThemeOption="DARK" />);
    const { getByTestId } = container;

    expect(getByTestId('1-DARK').props.checked).toStrictEqual(true);
    expect(getByTestId('0-LIGHT').props.checked).toStrictEqual(false);
    expect(getByTestId('2-SYSTEM').props.checked).toStrictEqual(false);
  });

  it('should call on onThemeChange callback when theme radio button clicked', () => {
    const container = render(<ThemesOptions {...props} />);
    const { getByTestId } = container;

    fireEvent(getByTestId('1-DARK'), 'onChange');

    expect(mockThemeChange).toHaveBeenCalledTimes(1);
    expect(mockThemeChange).toHaveBeenCalledWith('DARK');

    fireEvent(getByTestId('2-SYSTEM'), 'onChange');

    expect(mockThemeChange).toHaveBeenCalledWith('SYSTEM');
  });

  it('should not render system theme option if appearance value is no-preference', () => {
    jest.spyOn(Appearance, 'getColorScheme').mockReturnValue('no-preference');

    const container = render(<ThemesOptions {...props} />);
    const { queryByTestId } = container;

    expect(queryByTestId('2-SYSTEM')).toBeNull();
    expect(container).toMatchSnapshot();
  });
});
