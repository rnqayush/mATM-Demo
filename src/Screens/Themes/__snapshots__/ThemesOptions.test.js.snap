// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSettings screen should not render system theme option if appearance value is no-preference 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": 32,
        "paddingHorizontal": 16,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <SvgMock
      height={170}
      width={79}
    />
    <View
      checked={true}
      id={0}
      onChange={[Function]}
      style={
        Array [
          Object {
            "paddingTop": 16,
          },
        ]
      }
      testID="0-LIGHT"
      value="LIGHT"
    />
    <View
      align="center"
      fontSize={16}
      lineHeight={24}
    >
      Light
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <SvgMock
      height={170}
      width={79}
    />
    <View
      checked={false}
      id={1}
      onChange={[Function]}
      style={
        Array [
          Object {
            "paddingTop": 16,
          },
        ]
      }
      testID="1-DARK"
      value="DARK"
    />
    <View
      align="center"
      fontSize={16}
      lineHeight={24}
    >
      Dark
    </View>
  </View>
</View>
`;

exports[`ThemeSettings screen should render component 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": 32,
        "paddingHorizontal": 16,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <SvgMock
      height={170}
      width={79}
    />
    <View
      checked={true}
      id={0}
      onChange={[Function]}
      style={
        Array [
          Object {
            "paddingTop": 16,
          },
        ]
      }
      testID="0-LIGHT"
      value="LIGHT"
    />
    <View
      align="center"
      fontSize={16}
      lineHeight={24}
    >
      Light
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <SvgMock
      height={170}
      width={79}
    />
    <View
      checked={false}
      id={1}
      onChange={[Function]}
      style={
        Array [
          Object {
            "paddingTop": 16,
          },
        ]
      }
      testID="1-DARK"
      value="DARK"
    />
    <View
      align="center"
      fontSize={16}
      lineHeight={24}
    >
      Dark
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <SvgMock
      height={170}
      width={79}
    />
    <View
      checked={false}
      id={2}
      onChange={[Function]}
      style={
        Array [
          Object {
            "paddingTop": 16,
          },
        ]
      }
      testID="2-SYSTEM"
      value="SYSTEM"
    />
    <View
      align="center"
      fontSize={16}
      lineHeight={24}
    >
      System
    </View>
  </View>
</View>
`;
